### PR TITLE
[asft] feat: add Anchored Supervised Fine-Tuning (ASFT) recipe

### PR DESCRIPTION
**Pull request by zhuchichi56**
### What does this PR do?

Adds a new self-contained recipe `recipe/asft/` implementing **Anchored Supervised Fine-Tuning (ASFT)** — ICLR 2026, [arXiv:2509.23753](https://arxiv.org/abs/2509.23753) — under verl, alongside SFT and DFT baselines for fair comparison. ASFT extends *Dynamic Fine-Tuning (DFT)* with a KL anchor against a frozen reference model, recovering DFT's tighter RL lower bound while removing its distributional drift.

The recipe ships a forked `fsdp_sft_trainer_asft.py` exposing three loss modes (`sft / dft / asft`) plus an MCQ benchmark eval hook. **No changes are made to core verl** — the recipe follows the same convention as `dapo/`, `spin/`, `sppo/`, `gvpo/`: one folder, a `REQUIRED_VERL.txt` pin, its own trainer + launcher.

Reference project: https://github.com/zhuchichi56/ASFT

### Checklist Before Starting

- [x] Search for similar PRs: no prior ASFT recipe found in this repo (`grep -ri asft` returns only this PR's files).
- [x] PR title formatted as `[asft] feat: ...` (matches `recipe` module style; ASFT is the new subfolder name).

### Test

Validated with the bundled `run_asft.sh` 8-GPU launcher on Qwen2.5-7B-Base over the medical MCQ split of `chichi56/ASFT` (med: medqa / mmlu-medical / medmcqa). ASFT outperforms SFT and DFT consistently; the trainer mirrors the upstream verl 0.6.x FSDP SFT trainer plus the KL anchor term, so it inherits existing FSDP correctness. Reproduction details and ablations are in the upstream paper (Tables 1–3) and at https://github.com/zhuchichi56/ASFT.

### API and Usage Example

```bash
# 1. Prepare data (downloads chichi56/ASFT med split)
python recipe/asft/prepare_data.py --dataset med --output_dir ./data/asft_med

# 2. Train ASFT on 8 GPUs (sweep sft / dft / asft for fair comparison)
LOSS_MODES="sft dft asft" bash recipe/asft/run_asft.sh
```

Key config knobs added (all under `trainer.*`, opt-in only):

| key | default | meaning |
|-----|---------|---------|
| `loss_mode` | `sft` | one of `sft`, `dft`, `asft` |
| `asft_kl_coef` | `0.1` | β for the KL anchor (we recommend `0.03` under bf16) |
| `benchmark_eval_dir` | `null` | optional MCQ in-training eval dir |

### Design & Code Changes

- All new files live under `recipe/asft/`. **No core verl files are modified.**
- ASFT loss = DFT term `-E[ p_θ(y|x).detach() · log p_θ(y|x) ]` + `β · KL( p_θ ‖ p_ref )`, with `p_ref` a frozen FSDP-wrapped clone of the initial policy.
- Reference model is built once at startup (only when `loss_mode=asft`), avoiding any cost for the `sft` / `dft` paths.
- `sft_trainer.yaml` is a verbatim copy of the verl 0.6.x SFT config plus two opt-in keys (`loss_mode`, `asft_kl_coef`) — defaults preserve standard SFT behaviour.

### Checklist Before Submitting

- [ ] Read the Contribute Guide.
- [x] Apply `pre-commit` checks — will run before merge once reviewer confirms scope is acceptable.
- [ ] README under `recipe/asft/README.md` documents usage, config, and the design rationale.
- [x] CI tests — algorithmic recipe; correctness validated by paper/experiments rather than CI (matches convention used by other algorithm recipes such as `dapo/`, `sppo/`).
- [x] Recipe-only change; no submodule pin update needed beyond merging this PR into `verl-recipe`.

### CODE REVIEW

**Comment by zhuchichi56** [comment]
Thanks for the thorough review! Pushed `20ae58f` addressing the high-priority items:

- **[high] missing `importlib.util`** — added explicit `import importlib.util`.
- **[high] KL direction & stability** — switched to forward KL `KL(p_theta || p_ref)` via `F.kl_div(input=log p_ref, target=log p_theta, log_target=True)`, now matches the README and the paper.
- **[high] FSDP `.generate()`** — the eval path now calls `FSDP.summon_full_params(...)` and invokes `generate` on the underlying HF module instead of the FSDP wrapper.
- **[medium] default `fsdp2` vs ASFT-only-`fsdp`** — switched `sft_trainer.yaml` default to `strategy: fsdp` so users following the default no longer hit the guard; left a comment that `sft`/`dft` modes also work with `fsdp2`.
- **[medium] `vocab_size` reshape** — replaced with `shift_logits.size(-1)`.
- **pre-commit** — ran `ruff` + `ruff-format`; all checks pass locally.

The remaining medium items (DFT/ASFT under sequence parallelism, batched MCQ generation, more robust option-token tokenization) are intentional follow-up work — the current recipe reproduces the paper's exact setup and I'd prefer to land those as a focused follow-up PR rather than expanding the scope here. Happy to file tracking issues if helpful.

**Comment by zhuchichi56** [comment]
Quick correction on the KL-direction item — after re-checking the ASFT paper (arXiv:2509.23753, §4.2 Eq. 6 and §5 / Fig. 2):

> L_ASFT(θ) = L_DFT(θ) + λ · E_s[ D_KL( π_base(·|s) ‖ π_θ(·|s) ) ]

The paper explicitly **adopts the `P‖Q` convention with `P = π_base` (frozen reference) and `Q = π_θ` (policy)**, and **defines this direction as the *forward* KL**. Section 5 / Figure 2 ablate it against the mode-seeking *reverse* KL `D_KL(π_θ ‖ π_base)` and show forward KL wins — that's the central design choice of ASFT.

So the original implementation was actually correct; my previous commit (`20ae58f`) accidentally swapped the arguments. Just pushed `bfbaf8f` restoring `D_KL(p_ref ‖ p_θ)` (paper Eq. 6), kept `log_target=True` for numerical stability, and added an explicit comment about the convention so this doesn't trip up future readers.

Apologies for the back-and-forth, and thanks again for the careful review.
